sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,500 @@
+      return new BoundedMongoDbReader(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
+      long estimatedSizeBytes = 0L;
+
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+      MongoCollection mongoCollection = mongoDatabase.getCollection(collection);
+
+      // get the Mongo collStats object
+      // it gives the size for the entire collection
+      BasicDBObject stat = new BasicDBObject();
+      stat.append("collStats", collection);
+      Document stats = mongoDatabase.runCommand(stat);
+      estimatedSizeBytes = Long.valueOf(stats.get("size").toString());
+      long avgSize = Long.valueOf(stats.get("avgObjSize").toString());
+      if (filter != null && !filter.isEmpty()) {
+        // as the user is using a filter, we improve the estimated size
+        // by counting number of document returned by the filter
+        // and using the average object size from collStats
+        try {
+          estimatedSizeBytes = getFilterResultByteSize(mongoCollection, avgSize);
+        } catch (Exception e) {
+          LOGGER.warn("Can't estimate size", e);
+        }
+      }
+      return estimatedSizeBytes;
+    }
+
+    private long getFilterResultByteSize(MongoCollection mongoCollection, long avgSize)
+        throws Exception {
+      // count the number of documents selected by the filter and estimate the size of the
+      // filter using the average object size * the number of documents
+      long totalFilterByteSize = 0L;
+      Document bson = Document.parse(filter);
+      long countFilter = mongoCollection.count(bson);
+      totalFilterByteSize = countFilter * avgSize;
+      return totalFilterByteSize;
+    }
+
+    @Override
+    public List<BoundedSource> splitIntoBundles(long desiredBundleSizeBytes,
+                                                PipelineOptions options) {
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+
+      ArrayList splitKeys = null;
+      if (numSplits > 0) {
+        // the user defines his desired number of splits
+        // calculate the bundle size
+        long estimatedSizeBytes = getEstimatedSizeBytes(options);
+        desiredBundleSizeBytes = estimatedSizeBytes / numSplits;
+      }
+
+      // now we have the desized bundle size (provided by user or provided by the runner)
+      // we use Mongo splitVector command to get the split keys
+      BasicDBObject splitVectorCommand = new BasicDBObject();
+      splitVectorCommand.append("splitVector", database + "." + collection);
+      splitVectorCommand.append("keyPattern", new BasicDBObject().append("_id", 1));
+      splitVectorCommand.append("force", false);
+      // maxChunkSize is the Mongo partition size in MB
+      splitVectorCommand.append("maxChunkSize", desiredBundleSizeBytes / 1024 / 1024);
```

Comment:
* What will this do if desiredBundleSizeBytes is zero or small?
* Please add logging about how many parts you actually split into. It will be valuable for debugging when a pipeline is slow and it is because the source split into very few parts.